### PR TITLE
Adding version check for hypothesis deadline

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31262 Adding version check for hypothesis deadline**

Differential Revision: [D19036700](https://our.internmc.facebook.com/intern/diff/D19036700)

### CODE REVIEW

File: test/hypothesis_utils.py [hunk 1]
```diff
@@ -304,11 +305,18 @@ def tensor_conv(
...
+    settings.register_profile("no_deadline", deadline=None)
+
+    # This is really just to get flake8 to not complain when this file
+    # is imported purely for the side-effectful stuff above
+    def assert_deadline_disabled():
+        assert settings().deadline is None
+else:
+    from warnining import warn
+    warn("Your hypothesis version is outdated and doesn't support `deadline`. "
+         "Some tests are likely to fail with `DeadlineExceeded` error. "
```
Comment (jamesr66a) [inline test/hypothesis_utils.py]: This seems like a contradiction with the sentence before it

**Review by jamesr66a — COMMENTED** [review]
We should probably just require a recent Hypothesis version rather than adding complexity to our codebase. Is this just needed because of the default `hypothesis` version on devservers?

File: test/hypothesis_utils.py (hunk 1 above)
Comment (z-a-f) [reply test/hypothesis_utils.py]: Oh it does read like it. I should add that the `deadline` is disabled, and now the tests can fail with...

**Comment by z-a-f** [comment]
> We should probably just require a recent Hypothesis version rather than adding complexity to our codebase. Is this just needed because of the default `hypothesis` version on devservers?

Yes -- I cannot run most tests on devserver because of the versions

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit ad1d80f1:


* **1/1** broken upstream at merge base 68e51723 (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=68e517238291d1149a89d1f4e9fc90b560804d00&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **0/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=ad1d80f1e5a170785162ac590aea8f674728b25d).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_linux_xenial_py3_clang5_asan_test](https://dr.pytorch.org/build-details.html?build_id=78587548)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: test/hypothesis_utils.py [hunk 2]
```diff
@@ -304,11 +305,18 @@ def tensor_conv(
...
-settings.register_profile("no_deadline", deadline=None)
-settings.load_profile("no_deadline")
-
-# This is really just to get flake8 to not complain when this file
-# is imported purely for the side-effectful stuff above
+# We set the deadline in the currently loaded profile.
+# Creating (and loading) a separate profile overrides any settings the user
+# already specified.
+settings._profiles[settings._current_profile].__dict__['deadline'] = None
+settings._profiles[settings._current_profile].__dict__['timeout'] = hypothesis.unlimited
```
Comment (jerryzh168) [inline test/hypothesis_utils.py]: is `hypothesis.unlimited` valid for hypothesis version < 3.27.0?

Comment (z-a-f) [reply test/hypothesis_utils.py]: Good point!

**Review by jerryzh168 — APPROVED** [review]
LGTM

**Pull request closed without merging**